### PR TITLE
Implement parquet file appending.

### PR DESCRIPTION
**Pull request by yohplala**
(original text from original PR)

Hi @bmoscon ,

This PR implements appending of parquet files, along with a counter that keeps track the number of times each parquet file is incremented.
This feature solves problem presented in #19 (albeit being not exactly the feature requested) and problem presented in #108.
This branch has been made over PR #105, so commit from from PR #105 is already in this PR.

**Implementation note 1**
There is one counter per exchange/data_type/pair data for coping up with the case of scarcely updated data. In this case, the appending rate of each file is not necessarily the same (when there is no data for a exchange/data_type/pair, this specific data is not dumped into parquet file as per legacy implementation). Each parquet file follows then its own appending rate, so to say.
(I implemented this feature actually in 3 different versions this week-end, and earlier versions used a single counter for all files)

**Implementation note 2**
As indicated in config file, a not closed parquet file cannot be read. But a closed file cannot be appended efficiently: it needs to be read 1st and re-written and left open. Then new data can be appended.
So if cryptostore is stopped for any reason, last files are very likely unreadable (it requires to stop cryptostore at the last appending iteration when it is also closed).

If it seems to you required, I could implement in a new PR a new option to be activated by the user in the config file to enforce closing parquet file after each appending.
Reading and re-writing operations will certainly consume more CPU and memory however.

I decided not to retain this approach as a baseline, but to retain the most efficient way. Between each appending, file is left open, and only closed when the appending_counter is reached.

An alternative would be to run the closing task as soon as cryptostore exits. I do not know where to implement this actually?
If you give me a pointer and think it is appropriate, I can do that in a subsequent PR.

**Implementation note 3**
Well, I guess you already had this in mind, but if implementing multi-threading for data writing, please, keep this line out of the multi-threading stuff.

Bests

### CODE REVIEW

File: cryptostore/data/parquet.py [hunk 1]
```diff
@@ -21,7 +21,7 @@ class Parquet(Store):
 
     default_path = lambda self, exchange, data_type, pair: f'{exchange}/{data_type}/{pair}'
 
-    def __init__(self, exchanges, config=None):
+    def __init__(self, exchanges, buffer, config=None):
```
Comment (bmoscon) [inline cryptostore/data/parquet.py:24]: this shouldnt be a required arg, it should be optional, or in kwargs

File: cryptostore/data/storage.py [hunk 2]
```diff
@@ -12,17 +12,17 @@
 
 
 class Storage(Store):
-    def __init__(self, config):
+    def __init__(self, config, parquet_buffer):
```
Comment (bmoscon) [inline cryptostore/data/storage.py:15]: see previous comment. better to have kwargs and have the parquet specific implementation read that from out of it

**Review by bmoscon — CHANGES_REQUESTED** [review]

**Comment by bmoscon** [comment]
I'm not positive this is needed - it seems like that people want is something where the file is a specific size, or the file contains a specific amount of data (time-wise). I think this is a good start, but rather than have append N times wouldnt it be better to aim for one of those other metrics? I suppose if you care about time window size, it wouldnt be hard to do 1 minute updates and then close the file out after 60 of them

**Comment by bmoscon** [comment]
Also, can you confirm/verify that if the append counter is set to 1 that the current behavior remains the same? (So if its 1 it always closes the file after the first write, no appending is done?)

**Comment by yohplala** [comment]
> Also, can you confirm/verify that if the append counter is set to 1 that the current behavior remains the same? (So if its 1 it always closes the file after the first write, no appending is done?)

@bmoscon 
`append_counter` really is a counter for appending, it does not include the 1st writing step.
If you don't set it up in the config file, or if it is set to 0, there is no appending, and the file is closed juste after being written.

If it is set to 1, it will append once, so at a 1st iteration, it will write, then in a second, it will append, and close.

'if its 1 it always closes the file after the first write, no appending is done?'
No, `append_counter` set to 1 makes 1 appending: it is not what you have understood.

**Comment by yohplala** [comment]
> I'm not positive this is needed - it seems like that people want is something where the file is a specific size, or the file contains a specific amount of data (time-wise). I think this is a good start, but rather than have append N times wouldnt it be better to aim for one of those other metrics? I suppose if you care about time window size, it wouldnt be hard to do 1 minute updates and then close the file out after 60 of them

At 1st, I thought as you did: thinking 'time wise'. And after I realized that with this 'simple' approach, time wise content of the parquet file is ensured. Indeed the counter gives following specific amount of data (time-wise).:
     storage_interval x (append_counter +1)

This is because it is N times the `storage_interval`.

**Comment by yohplala** [comment]
Myself, I have finally set:
`storage_interval`: 6M
`append_counter`: 19
-> my parquet files contain 2 hours of data before being uploaded to Google Drive.

**Comment by yohplala** [comment]
The funny fact is that you state it yourself in your comment :) 
'1 minute updates and then close the file out after 60 of them'

This is exactly what it does:
1mn is `storage_interval`
60 is the `append_counter` (well actually, this would be 59)

File: cryptostore/data/storage.py (hunk 2 above)
Comment (yohplala) [reply cryptostore/data/storage.py]: @bmoscon : done

File: cryptostore/data/parquet.py (hunk 1 above)
Comment (yohplala) [reply cryptostore/data/parquet.py]: @bmoscon : done

**Review by yohplala — COMMENTED** [review]
@bmoscon 
Hi Bryant, I changed the code following your feedback. Please, let me know if there is anything else to modify.
Have a good day,
Bests

File: config.yaml [hunk 3]
```diff
@@ -91,9 +91,13 @@ influx:
...
 # Parquet specific options. Parquet will default to storing the data on disk unless these are specified
 parquet:
-    # if storing the data to an external source (like S3) toggle this to enable the removal of the local file after
+    # If storing the data to an external source (like S3) toggle this to enable the removal of the local file after
     # writing to external store
     del_file: true
+    # Append X times a parquet file before creating a new one to store data.
+    # This feature allows dumping frequently data from Redis, while writing less frequently new parquet files.
+    # Beware that a parquet file not properly closed cannot be read back. Closing is made at the last appending iteration.
+    append_counter: 4
```
Comment (bmoscon) [inline config.yaml:100]: Can you just add something here that this can be commented out or removed, or set to 0 to disable the append behavior?

**Review by bmoscon — CHANGES_REQUESTED** [review]

**Comment by bmoscon** [comment]
just one last change!

File: config.yaml (hunk 3 above)
Comment (yohplala) [reply config.yaml:101]: done

**Comment by yohplala** [comment]
> just one last change!

Hi, hi cap'tain!
Should be ok now.
:)

**Review by bmoscon — APPROVED** [review]

**Pull request merged**